metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java

```diff
@@ -442,14 +442,16 @@ SnapshotManifest loadSnapshot(
         while (reader.hasNext()) {
             Batch<ApiMessageAndVersion> batch = reader.next();
             loadControlRecords(batch);
-            for (ApiMessageAndVersion record : batch.records()) {
-                try {
-                    delta.replay(record.message());
-                } catch (Throwable e) {
-                    faultHandler.handleFault("Error loading metadata log record " + snapshotIndex +
-                            " in snapshot at offset " + reader.lastContainedLogOffset(), e);
+            if (reader.snapshotId().equals(Snapshots.BOOTSTRAP_SNAPSHOT_ID) && !batch.records().isEmpty()) {
+                for (ApiMessageAndVersion record : batch.records()) {
+                    try {
+                        delta.replay(record.message());
+                    } catch (Throwable e) {
+                        faultHandler.handleFault("Error loading metadata log record " + snapshotIndex +
+                                " in snapshot at offset " + reader.lastContainedLogOffset(), e);
+                    }
+                    snapshotIndex++;
```

Comment:
This is incorrect because the code is replaying records ONLY if the snapshot being loaded is the `0-0.checkpoint` with metadata records. The code should skip over replaying records in that case, not replay them. For all other checkpoints, the code should replay records.

```suggestion
            if (!reader.snapshotId().equals(Snapshots.BOOTSTRAP_SNAPSHOT_ID)) {
                for (ApiMessageAndVersion record : batch.records()) {
                    try {
                        delta.replay(record.message());
                    } catch (Throwable e) {
                        faultHandler.handleFault("Error loading metadata log record " + snapshotIndex +
                                " in snapshot at offset " + reader.lastContainedLogOffset(), e);
                    }
                    snapshotIndex++;
```